rama-tcp/src/client/pool.rs

```diff
@@ -0,0 +1,150 @@
+use std::sync::{atomic::AtomicUsize, Arc};
+
+use rand::seq::SliceRandom;
+
+use super::TcpStreamConnector;
+
+/// A pool of TcpConnectors
+#[derive(Debug, Clone)]
+pub struct TcpStreamConnectorPool<C> {
+    idx: Arc<AtomicUsize>,
+    connectors: Vec<C>,
+}
+
+impl<C: TcpStreamConnector> TcpStreamConnectorPool<C> {
+    /// A `TcpStreamConnector` where each connection is chosed randomly from a pool of
+    /// `TcpStreamConnector`s
+    pub fn new_random(connectors: Vec<C>) -> Self {
+        Self::new_random_with_rng(connectors, rand::rng())
+    }
+
+    fn shuffle_connectors<Rng: rand::Rng>(mut connectors: Vec<C>, mut rng: Rng) -> Vec<C> {
+        connectors.shuffle(&mut rng);
+        connectors
+    }
+
+    /// A `TcpStreamConnector` where each connection is chosed randomly from a pool of
+    /// `TcpStreamConnector`s using the specified `rand::Rng`
+    pub fn new_random_with_rng<Rng: rand::Rng>(connectors: Vec<C>, rng: Rng) -> Self {
+        Self {
+            idx: Arc::new(AtomicUsize::default()),
+            connectors: Self::shuffle_connectors(connectors, rng),
+        }
+    }
+
+    /// New 'Round Robin' `TcpStreamConnector`
+    pub fn new_round_robin(connectors: Vec<C>) -> Self {
+        Self {
+            idx: Arc::new(AtomicUsize::default()),
```

Comment:
I would randomize the index at creation time so you don't always start with the same connector, WDYT?

Reply:
The reason I provide this method is so I can build unit tests -- however in this case I think there are two tests:

1. Does the impl provide a random selection from a set of connectors
2. Does the impl return more connectors than the original set

Reply:
In the end I left it for the user to decide. If they want a randomised index they can pass a shuffled `Vec`

Reply:
Agreed, elegant and better. Thanks for standing up :)